sentry-android-replay/src/main/java/io/sentry/android/replay/viewhierarchy/ViewHierarchyNode.kt

```diff
@@ -0,0 +1,173 @@
+            val blue = Color.blue(color)
+            return Color.argb(alpha, red, green, blue)
+        }
+
+        fun fromView(view: View): ViewHierarchyNode {
+            // TODO: Extract redacting into its own class/function
+            // TODO: extract redacting into a separate thread?
+            var shouldRedact = false
+            var dominantColor: Int? = null
+            var rect: Rect? = null
+            when (view) {
+                is TextView -> {
+                    // TODO: API level check
+                    // TODO: perhaps this is heavy, might reconsider
+                    val nodeInfo = if (VERSION.SDK_INT >= VERSION_CODES.R) {
+                        AccessibilityNodeInfo()
+                    } else {
+                        AccessibilityNodeInfo.obtain()
+                    }
+                    view.onInitializeAccessibilityNodeInfo(nodeInfo)
+                    shouldRedact = nodeInfo.isVisibleToUser
+                    nodeInfo.recycle()
+                    if (shouldRedact) {
+                        val bounds = Rect()
+                        val text = view.text.toString()
+                        view.paint.getTextBounds(text, 0, text.length, bounds)
+                        dominantColor = adjustAlpha(view.currentTextColor)
+                        rect = Rect()
+                        view.getGlobalVisibleRect(rect)
+
+                        var textEnd = Int.MIN_VALUE
+                        var textStart = Int.MAX_VALUE
+                        if (view.layout != null) {
+                            for (i in 0 until view.layout.lineCount) {
+                                val min = view.layout.getLineStart(i)
+                                val minPosition = view.layout.getPrimaryHorizontal(min).toInt()
+                                val max = view.layout.getLineVisibleEnd(i)
+                                val maxPosition = view.layout.getPrimaryHorizontal(max).toInt()
+                                if (minPosition < textStart) {
+                                    textStart = minPosition
+                                }
+                                if (maxPosition > textEnd) {
+                                    textEnd = maxPosition
+                                }
+                            }
+                        } else {
+                            textEnd = rect.right - rect.left
+                            textStart = 0
+                        }
+                        // TODO: support known 3rd-party widgets like MaterialButton with an icon
+                        // TODO: also calculate height properly based on text bounds
+                        rect.left += textStart + view.paddingStart
+                        rect.right = rect.left + (textEnd - textStart)
+                    }
+                }
+
+                is ImageView -> {
+                    shouldRedact = isVisible(view) && (view.drawable?.isRedactable() ?: false)
```

Comment:
m: Can we check whether the image belongs to the APK or it was loaded from outside?

Reply:
no, there's no really a way to know this for sure, but we can do heuristics like seen in the `isRedactable` method - if it's not a `BitmapDrawable` then most likely it's coming from the APK (like icons or backgrounds). But if it's a BItmapDrawable then there's a chance we'll redact something that is bundled with the APK, but I think it's ok. We could do something at build time to preserve this information via the gradle plugin, but sounds like an overkill

Reply:
Im just worried avout redacting big portions based on a image background.

Reply:
yeah, makes sense 👍 I'll leave a TODO to see if we can do anything about that